File: Java/InsertionSort.java
import java.util.*;
import java.time.Duration;
import java.time.Instant;

public class InsertionSort {


    // Function to measure memory usage returns usage in kb
    static long getMemoryUsage() {
        Runtime runtime = Runtime.getRuntime();
        runtime.gc();
        return (runtime.totalMemory() - runtime.freeMemory());
    }

    static void insertionSort(ArrayList<Integer> arr) {
        int size = arr.size();
        for (int i = 1; i < size; i++) {
            int current = arr.get(i);
            int temp = i - 1;
            while (temp >= 0 && arr.get(temp) > current) {
                arr.set(temp + 1, arr.get(temp));
                temp--;
            }
            arr.set(temp + 1, current);
        }
    }

    public static void main(String[] args) {

        Scanner scanner = new Scanner(System.in);

        System.out.print("Enter a size: ");
        int size = scanner.nextInt();

        ArrayList<Integer> array = new ArrayList<>();
        Random rand = new Random();
        for (int i = 0; i < size; i++) {
            array.add(rand.nextInt(100));


        }

        long start = System.nanoTime();
        

        // function call
        insertionSort(array);
        long stop = System.nanoTime();
        
        long duration = stop - start;
        

       
        System.out.println("Insertion sort time is: " + (duration / 1000) + " microseconds");
       
        System.out.println("Memory usage: " + getMemoryUsage() + " KB");
       

        // prints the list to make sure the algorithm sorts the list properly
        for (int j = 0; j < size; j++) {
            System.out.print(" " + array.get(j));
        }

        System.out.println();
    }
}

File: C++ Code/For mac/insertionSort.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <sys/types.h>
#include <sys/sysctl.h>

using namespace std::chrono;
using namespace std;

// Function to measure memory usage returns usage in kilobytes
long getMemoryUsage() {
    struct rusage rusage;
    getrusage(RUSAGE_SELF, &rusage);
    return rusage.ru_maxrss;
}

// Function to measure CPU usage (macOS specific)
double getCPUUsage() {
    size_t size = sizeof(int);
    int cpuPercent;
    sysctlbyname("kern.cp_time", &cpuPercent, &size, NULL, 0);
    return cpuPercent / 100.0;
}

void insertionSort(std::vector<int>& arr, int size) {
    for (int i = 1; i < size; i++) {
        int current = arr[i];
        int temp = i - 1;
        while (temp >= 0 && arr[temp] > current) {
            arr[temp + 1] = arr[temp];
            temp--;
        }
        arr[temp + 1] = current;
    }
}

int main() 
{
    // get random sample size
    vector<int> array;
    int size;
    cout << "Enter a size: ";
    cin >> size;
    
    for (int i = 0; i < size; i++)
    {
        array.push_back(rand() % 100);
        array.push_back(array[i]);
    }

    // Getting the time of competition for the execution, CPU, and memory usage
    auto start = high_resolution_clock::now();
   
    insertionSort(array, size);
    auto stop = high_resolution_clock::now();
  
    auto duration = duration_cast<microseconds>(stop - start);

    // prints the statements out
    cout << "Insertion sort time is: " << duration.count() << " microseconds" << endl;
    cout << "CPU usage: " << getCPUUsage() * 100 << "%" << endl; // CPU usage in percentage
    cout << "Memory usage: " << getMemoryUsage() << " KB" << endl;

    // prints the list to make sure the algorithm sorts the list properly
    for (int j = 0; j < size; j++) {
        cout << ' ' << array[j];
    }

    cout << endl;
    return 0;
}
